### ISSUE TITLE
Cryptographic signing of platform releases

### ISSUE THREAD

**Issue by palmskog** [coq/platform#132]
To increase the trustworthiness of the platform and allow users to certify that they are installing what they expect, I suggested in [a Coq Zulip topic](https://coq.gitlab.io/zulip-archive/stream/250632-Coq-Platform-devs-&-users/topic/Certifying.20platform.20release.20packages.20by.20signature.html) that platform releases should be cryptographically signed by the release manager.

There is a wide spectrum of possible ways to sign releases, going from signing of the GitHub commit for the platform repo associated with a release, to full signing of a special platform opam repository. Fortunately, the latter (arguably more fundamental) approach has been explored at length by @hannesm in his [Conex](https://github.com/hannesm/conex) project.

> Conex is a utility for verify and attest release integrity and authenticity of community repositories through the use of cryptographic signatures

The ideas behind Conex have been described in blog posts and papers (in reverse chronological order):
- https://hannes.nqsb.io/Posts/Conex
- https://github.com/hannesm/conex-paper/raw/master/paper.pdf
- https://opam.ocaml.org/blog/Signing-the-opam-repository/

According to Hannes, Conex is as of September 2021 not yet ready for deployment at the scale of the Coq platform, but is expected to improve later this year, which could make it viable to use here.

Hannes summarizes the design of Conex as an adaptation of the [Update framework](https://theupdateframework.io) for opam:

> the signing metadata is part of the opam repository --> an update of the opam repository can be verified (and this is the only place where verification needs to happen)

> conex has few dependencies, and will likely be integrated into opam at some point in time

> opam has (since 2.0) a so called validation hook https://opam.ocaml.org/doc/Manual.html#configfield-repository-validation -> command which is executed after repository update (but before installing the update of the repository) with the old repo (on disk) and the patch to be applied --> here conex can verify the new repository to be properly signed

**Comment by MSoegtropIMC** [comment]
@palmskog : do you think singing the github commits (annotated tags) would be sufficient for the time being?

**Comment by palmskog** [comment]
@MSoegtropIMC annotated tags sound good, but for this to be meaningful I think signing should be done with a personal GPG key that can be independently checked to belong to one's web of trust. For example, I documented how one could recreate/check a Coq tag/release [on the Coq Wiki](https://github.com/coq/coq/wiki/VerifyingReleases).

**Comment by MSoegtropIMC** [comment]
Wouldn't simply be signing the tag with a GPG key that is registered with Github sufficient?

**Comment by MSoegtropIMC** [comment]
Yes, I can do this key signing with @gares or ask the respective Coq release manager to sign of the Coq Platform tag.

**Comment by gares** [comment]
Didn't we sign our GPG keys already?

**Comment by gares** [comment]
> Maybe we can ask @gares to sign the signing key, or to sign the commit directly? Then you at least have the connection between Coq release signing and Platform signing.

I think all devs should cross sign their keys, then `git tag -s` should just be enough. My key is decently connected (Debian past).

**Comment by MSoegtropIMC** [comment]
@gares: we did exchange the keys but I don't think that either of use signed the signing keys of the others.

**Comment by MSoegtropIMC** [comment]
Triage note: all future tags will be signed with the personal GPG key of the support engineer of Coq Platform at INRIA - currently @rtetley. @palmskog : if this is agreeable, please close this issue.

**Comment by MSoegtropIMC** [comment]
(Or close after it actually happened for the next release)

**Comment by palmskog** [comment]
@MSoegtropIMC signing the tag is one big step forward, but as per earlier discussion in this issue I don't think it solves verifiability by Platform users of the package collection, which something like [conex](https://github.com/hannesm/conex) would solve.

**Comment by MSoegtropIMC** [comment]
In this case I think the Coq team needs to state what kind of credibility authority or mechanism INRIA wants to use.

@Zimmi48 : can you please comment?

**Comment by palmskog** [comment]
Since @Zimmi48 is not an Inria employee, so I think it's hard for him to comment on Inria issues in general. But arguably, there needs to be some kind of authority backing up all signing related to Coq and the Platform. Ideally, there is a trace of cryptographic signings going from the Platform maintainer to a Coq core team coordinator (and possibly further into Inria-based authority). I don't see how we can have formal trust otherwise. There are already [examples](https://en.wikipedia.org/wiki/XZ_Utils_backdoor) of open source projects being "taken over" and used as attack vectors.

**Comment by MSoegtropIMC** [comment]
Looking at how difficult it is to get the installers signed, I have little hope that we get anywhere any time soon with deriving something from INRIA owned keys.
